### ISSUE TITLE
Maximum Recursion Depth Reached Error

### ISSUE THREAD

**Issue by pgunn** [flatironinstitute/CaImAn#1204]
### Discussed in https://github.com/flatironinstitute/CaImAn/discussions/1201

<div type='discussions-op-text'>

<sup>Originally posted by **charan1319** October 19, 2023</sup>
Hello,

I'm trying to run the fit function on my cnmf object with 3D data but am getting a maximum recursion depth error. Is this simply because of the length of the recording? Either way, does anyone have suggestions for the best solution to this problem? Also, I would like to check the locations of cells that CaImAn found by plotting them on top of a frame of my recording but have been having some trouble getting this to work -- how can I do this? 

Thank you!

Code: 
<img width="665" alt="image" src="https://github.com/flatironinstitute/CaImAn/assets/37819328/72615f7a-541d-4b94-b1d2-03139225adca">

Error: 
<img width="1121" alt="image" src="https://github.com/flatironinstitute/CaImAn/assets/37819328/fd941803-2938-43bf-acbe-16b7460dd1e1">

</div>

**Comment by pgunn** [comment]
@charan1319 It looks like there may have been some cut text from the Error; is any of the error message "clicky" in that it'd let you expand it to see more? In particular, that ... between `return self` (near the bottom) and the paragraph after?

**Comment by charan1319** [comment]
@pgunn Yes, here's the screenshot after expanding it: 

<img width="1366" alt="image" src="https://github.com/flatironinstitute/CaImAn/assets/37819328/ae5e3ca0-1650-4edb-87df-c78b0ebaf678">
<img width="1363" alt="image" src="https://github.com/flatironinstitute/CaImAn/assets/37819328/34d2643a-5478-47c9-a985-1221ec13b0f4">
<img width="1090" alt="image" src="https://github.com/flatironinstitute/CaImAn/assets/37819328/0ec23c7f-5f41-415e-8d04-74e1f66199e9">
<img width="1118" alt="image" src="https://github.com/flatironinstitute/CaImAn/assets/37819328/4fe02efd-b7ac-463b-bd09-dd4e1b994b1a">


It looks like something else is causing an error in scipy/sparse/_dia.py and then the recursion error comes from the code being unable to figure out how to display the error? But I haven't been able to figure out what's causing the error in scipy in the first place.

**Comment by pgunn** [comment]
Hello,
Can you send us a copy of your testing.ipynb? If you like, send it via email to

pgunn@flatironinstitute.org

I haven't seen this kind of thing before.

**Comment by charan1319** [comment]
Just shared it. Thank you for your help!

**Comment by pgunn** [comment]
Hi,
Looking at the notebook, it looks like you hit an error in the cell above what produced the recursion depth message, but then you commented out the cell? It's hard to follow exactly what happened.

You might want to do "Kernel->Restart and Clear Output", and then carefully step through the cells, executing just the ones you want to execute and stopping the first time you hit anything that looks like an error (any cells after you get an error will likely fail because they tend to build on each other's results and assume things went well).

I am also curious about the dataset you're working on; it's reporting more dimensionality than the code expects. How big is that h5 file? If it's not huge, maybe we can take a look at it as a hint as to what's going on (we may also be able to help you with parameters to get the best results)

(have you run through the demo notebooks with our sample data successfully?)

**Comment by pgunn** [comment]
I just gave the notebook a more careful look; I think for 3d data you're probably going to have a better time starting with one of the two 3d notebooks we provide (demo_caiman_cnmf_3D or demo_online_3D).

**Comment by charan1319** [comment]
Yes, the error from the cell above was from running it set to a 2D input instead of a 3D, but the cell with the error I was referencing was indeed for 3D data. I also tried running it again from scratch with the 3D notebook and I get an estimates.A sparse vector that is 300x0 for some reason. And plotting the max projections also results in something that doesn't look right. Could you please let me know what might cause this? 

<img width="602" alt="image" src="https://github.com/flatironinstitute/CaImAn/assets/37819328/60e7a6d7-43b4-4d39-b59f-0bb09cd34db8">


<img width="508" alt="image" src="https://github.com/flatironinstitute/CaImAn/assets/37819328/e4785adc-37f7-40b0-873f-635d7ad84f6e">

**Comment by pgunn** [comment]
Hello, sorry for not getting back to you sooner on this; I got confused between this and someone else whom I was helping via email.

I suggest starting from one of the 3d notebooks with your file, because it will include all the specifics needed for 3d handling that might be missed by freely mixing things between; can you give that a go? If that still doesn't work we can dig into this further.

**Comment by charan1319** [comment]
Hello, not worries! I did try it from scratch with the demo_caiman_cnmf_3D notebook and reordering the dimensions of the data helped fix the projection issue, but I'm still getting 0 elements in estimates.A. Do you have any thoughts on why this might be happening? Here's an error I get when trying to view the components after running cnm.fit(): 
<img width="814" alt="image" src="https://github.com/flatironinstitute/CaImAn/assets/37819328/f01c84a3-17fe-4e81-a673-11ab7818451d">

**Comment by EricThomson** [comment]
Could you check that estimates is actually empty? I.e., what is `cnm.estimates.A.shape`? My _guess_ is this is a plotter issue not an estimates issue, but I could be wrong.

**Comment by EricThomson** [comment]
Closing lack of activity, please reopen if you still need help with this.
